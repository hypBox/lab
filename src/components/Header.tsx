import React from "react"
import {
  AppBar,
  Toolbar,
  IconButton,
  Typography,
  MenuItem,
  Menu,
} from "./material/core"
import config from "../utility/config"

import { User, Menu as MenuIcon, FAIcon } from "./material/icons"

import { makeStyles, createStyles } from "./material/styles"

const useStyles = makeStyles((theme) =>
  createStyles({
    root: {
      flexGrow: 1,
    },
    menuButton: {
      marginRight: theme.spacing(2),
    },
    title: {
      flexGrow: 1,
    },
  })
)

const Header = () => {
  const classes = useStyles()
  const { app: appConfig } = config
  const [anchorEl, setAnchorEl] = React.useState<null | HTMLElement>(null)
  const open = Boolean(anchorEl)

  const handleMenu = (event: React.MouseEvent<HTMLElement>) => {
    setAnchorEl(event.currentTarget)
  }

  const handleClose = () => {
    setAnchorEl(null)
  }

  return (
    <div className={classes.root}>
      <AppBar position="static">
        <Toolbar>
          <IconButton
            edge="start"
            className={classes.menuButton}
            color="inherit"
            aria-label="menu"
          >
            <FAIcon icon={MenuIcon} />
          </IconButton>
          <Typography variant="h6" className={classes.title}>
            {appConfig.title}
          </Typography>

          <div>
            <IconButton
              aria-label="account of current user"
              aria-controls="menu-appbar"
              aria-haspopup="true"
              onClick={handleMenu}
              color="inherit"
            >
              <FAIcon icon={User} />
            </IconButton>
            <Menu
              id="menu-appbar"
              anchorEl={anchorEl}
              anchorOrigin={{
                vertical: "top",
                horizontal: "right",
              }}
              // keepMounted
              transformOrigin={{
                vertical: "top",
                horizontal: "right",
              }}
              open={open}
              onClose={handleClose}
            >
              <MenuItem onClick={handleClose}>My Lab</MenuItem>
              <MenuItem onClick={handleClose}>Profile</MenuItem>
              <MenuItem onClick={handleClose}>Sign out</MenuItem>
            </Menu>
          </div>
        </Toolbar>
      </AppBar>
    </div>
  )
}

export default Header
