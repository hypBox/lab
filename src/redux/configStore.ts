import { createStore, Store, applyMiddleware } from "redux"
import rootReducer from "./rootReducer"

import { createLogger } from "redux-logger"
import { composeWithDevTools } from "redux-devtools-extension"
import reduxImmutableStateInvariant from "redux-immutable-state-invariant"

export interface AppState {
  preferences: {
    darkMode: boolean
  }
}

const middlewares: Array<any> = [reduxImmutableStateInvariant()]

if (process.env.NODE_ENV === "development") {
  middlewares.push(
    createLogger({
      collapsed: true,
    })
  )
}

const configStore = (): Store<AppState> =>
  createStore(rootReducer, composeWithDevTools(applyMiddleware(...middlewares)))

export default configStore
