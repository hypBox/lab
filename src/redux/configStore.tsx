import { createStore } from "redux"
import rootReducer from "./reducers"

export interface Store {
  darkTheme: boolean
}

const configStore = (intialState: any) => createStore(rootReducer, intialState)

export default configStore
