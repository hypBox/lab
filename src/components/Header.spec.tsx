import React from "react"
import { render, fireEvent, cleanup } from "@testing-library/react"
import { Provider as ReduxProvider } from "react-redux"
import configStore from "../redux/configStore"
import Header from "./Header"

const store = configStore()

afterEach(() => {
  cleanup()
})

describe("Header", () => {
  const renderHeader = () =>
    render(
      <ReduxProvider store={store}>
        <Header />
      </ReduxProvider>
    )

  it("should user menu items should not be shown by default", () => {
    const tree = renderHeader()
    expect(tree.queryByLabelText("menu-appbar")).toBeFalsy()
  })

  it("should toggle current user menu when menu is clicked twice", async () => {
    const tree = renderHeader()
    const userMenuButton = tree.getByLabelText("current-user")
    fireEvent.click(userMenuButton)
    expect(tree.findByLabelText("menu-appbar")).toBeTruthy()
    fireEvent.click(tree.getByText("My Lab"))
    expect(tree.queryByLabelText("menu-appbar")).toBeFalsy()
  })

  it("should display light theme by default", () => {
    const tree = renderHeader()
    expect(tree.getByTestId("enable-dark-theme")).toBeTruthy()
    expect(tree.queryByTestId("enable-light-theme")).toBeFalsy()
  })

  it("should change theme to dark", () => {
    const tree = renderHeader()
    const toggleThemeButton = tree.getByTestId("enable-dark-theme")
    fireEvent.click(toggleThemeButton)
    expect(tree.queryByTestId("enable-dark-theme")).toBeFalsy()
    expect(tree.getByTestId("enable-light-theme")).toBeTruthy()
  })
})
