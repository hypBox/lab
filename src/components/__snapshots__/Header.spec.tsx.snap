// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header renders as expected 1`] = `
<div
  className="makeStyles-root-1"
>
  <header
    className="MuiPaper-root MuiAppBar-root MuiAppBar-positionStatic MuiAppBar-colorPrimary MuiPaper-elevation4"
  >
    <div
      className="MuiToolbar-root MuiToolbar-regular MuiToolbar-gutters"
    >
      <button
        aria-label="menu"
        className="MuiButtonBase-root MuiIconButton-root makeStyles-menuButton-2 MuiIconButton-colorInherit MuiIconButton-edgeStart"
        disabled={false}
        onBlur={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="button"
      >
        <span
          className="MuiIconButton-label"
        >
          <svg
            aria-hidden="true"
            className="svg-inline--fa fa-bars fa-w-14 "
            data-icon="bars"
            data-prefix="fas"
            focusable="false"
            role="img"
            style={Object {}}
            viewBox="0 0 448 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M16 132h416c8.837 0 16-7.163 16-16V76c0-8.837-7.163-16-16-16H16C7.163 60 0 67.163 0 76v40c0 8.837 7.163 16 16 16zm0 160h416c8.837 0 16-7.163 16-16v-40c0-8.837-7.163-16-16-16H16c-8.837 0-16 7.163-16 16v40c0 8.837 7.163 16 16 16zm0 160h416c8.837 0 16-7.163 16-16v-40c0-8.837-7.163-16-16-16H16c-8.837 0-16 7.163-16 16v40c0 8.837 7.163 16 16 16z"
              fill="currentColor"
              style={Object {}}
            />
          </svg>
        </span>
      </button>
      <h6
        className="MuiTypography-root makeStyles-title-3 MuiTypography-h6"
      >
        hypBox Lab
      </h6>
      <div>
        <button
          aria-controls="menu-appbar"
          aria-haspopup="true"
          aria-label="account of current user"
          className="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit"
          disabled={false}
          onBlur={[Function]}
          onClick={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={0}
          type="button"
        >
          <span
            className="MuiIconButton-label"
          >
            <svg
              aria-hidden="true"
              className="svg-inline--fa fa-user fa-w-14 "
              data-icon="user"
              data-prefix="fas"
              focusable="false"
              role="img"
              style={Object {}}
              viewBox="0 0 448 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M224 256c70.7 0 128-57.3 128-128S294.7 0 224 0 96 57.3 96 128s57.3 128 128 128zm89.6 32h-16.7c-22.2 10.2-46.9 16-72.9 16s-50.6-5.8-72.9-16h-16.7C60.2 288 0 348.2 0 422.4V464c0 26.5 21.5 48 48 48h352c26.5 0 48-21.5 48-48v-41.6c0-74.2-60.2-134.4-134.4-134.4z"
                fill="currentColor"
                style={Object {}}
              />
            </svg>
          </span>
        </button>
      </div>
    </div>
  </header>
</div>
`;
